Support factory-created solution in AH boundary conditions, prepare for SHK initial data

nilsvu opened the pull request:
## Proposed changes

Use analytic solution base class in `Xcts::BoundaryConditions::ApparentHorizon`. The `Xcts::AnalyticData::Binary` already supports factory creation. Therefore, this change should enable SKS, SHK, and also mixed superposed KerrSchild/Harmonic coordinates (which I think haven't been tried in SpEC yet?), once a Kerr solution in those coords is implemented.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Fix a missing include

nilsvu committed:
Add a missing link lib

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp:
```diff
@@ -214,7 +215,8 @@ namespace Solutions {
  * wish to use the other transformation in the future.
  */
-class KerrSchild : public AnalyticSolution<3_st>,
+class KerrSchild : public Kerr,
```
@nikwit happy now? ;)

nilsdeppe commented on src/PointwiseFunctions/InitialDataUtilities/InitialData.hpp:
```diff
@@ -18,7 +18,7 @@ namespace initial_data {
  * analytic solutions and analytic data must virtually inherit from this class.
  */
-class InitialData : public PUP::able {
+class InitialData : public virtual PUP::able {
```
This seems like a really bad idea. Who's responsible for calling the base class `pup` function when you have multiple in the hierarchy? I don't think calling the function more than once is a good idea (seems like asking for UB). I'm really strongly opposed to this change unless there is clear documentation from Charm++ that this is supported and how to handle it correctly. This is not something I think we should leave to "the tests seem to pass so it must be fine". It could result in horrendous runtime errors.

nilsvu committed:
Add virtual get_clone to elliptic solutions

nilsvu committed:
Add ADM variables to XCTS solutions, tests

Allow retrieving spatial metric and extrinsic curvature as well.
Test consistency of a bunch of variables and ADM constraints.

nilsvu committed:
Support factory-created solution in AH boundary conditions

nilsvu committed:
Instantiate partial_derivative for another tensor type

nilsvu commented:
I rewrote this quite substantially, removing the Kerr base class. Instead, the AH boundary conditions work with the analytic solution base class now.

nilsdeppe commented on src/Elliptic/Systems/Xcts/BoundaryConditions/ApparentHorizon.cpp:
```diff
@@ -16,55 +16,34 @@
+        solution_for_lapse,
+    std::optional<std::unique_ptr<elliptic::analytic_data::AnalyticSolution>>
+        solution_for_negative_expansion,
+    const Options::Context& /*context*/)
```
can you remove this argument now?

nilsvu replied:
Yea I figured I'd leave it to make adding parse checks later easier 🤷‍♂️ no strong opinion here

nilsdeppe replied:
Sounds good! Doesn't matter to me :)

nilsdeppe reviewed (dismissed):
LGTM, one small thing that you can change if you want and it's okay

nilsvu committed:
Test XCTS solutions yet another way

nilsvu commented:
pushed another two commits with more tests and a small refactor

nilsvu committed:
Use partial_derivative function in XCTS computer

nilsvu committed:
Add Xcts::extrinsic_curvature function

Computes the extrinsic curvature from its conformal decomposition.

nilsvu commented:
Rebased on #3853, which required that I add the extrinsic curvature in the two added Schwarzschild coordinates. Did this in a fixup. Only the two last commits are new (the fixup and "Add Xcts::extrinsic_curvature function").

nilsdeppe reviewed (commented):
LGTM, please go ahead and squash. Thanks for pushing the extra commits. Made it a lot easier to review :)

nilsvu commented:
Rebased and increased the timeout of the Schwarzschild test a bit

geoffrey4444 reviewed (approved):
LGTM
